### PR TITLE
Upgraded extract_images_sync script to ROS 2

### LINKED ISSUE #860: Port extract_images_sync to ROS 2
The file image_view/scripts/extract_images_sync has not been ported to ROS 2

### PR DESCRIPTION
**Pull request by artineering**
Change Description: The extract_images_sync python script was upgraded to ROS 2.
Testing: Upgraded node tested against Iron on Ubuntu 22.04.3 LTS
Issue: fixes #860

### CODE REVIEW

File: image_view/package.xml
Comment (mikeferguson) [inline image_view/package.xml]: This could actually be an exec_depend since we don't have to build anything with this

File: image_view/package.xml
Comment (mikeferguson) [inline image_view/package.xml]: This can be reverted - maintainer is different than contributor - it's an ongoing commitment (and usually involves commit rights, reviewing PRs and releasing updates)

File: image_view/scripts/extract_images_sync
Comment (mikeferguson) [inline image_view/scripts/extract_images_sync]: Why the leading underscore in parameter name?

File: image_view/scripts/extract_images_sync
Comment (mikeferguson) [inline image_view/scripts/extract_images_sync]: These aren't quite right - check out how to use --ros-args (this would actually be interpreted as a remap as written)

File: image_view/scripts/extract_images_sync
Comment (mikeferguson) [inline image_view/scripts/extract_images_sync]: Param part is not correct - needs --ros-args and -p

File: image_view/scripts/extract_images_sync
Comment (mikeferguson) [inline image_view/scripts/extract_images_sync]: I think this logic is backwards right? Should not have "not"

**Review by mikeferguson — CHANGES_REQUESTED** [review]
Mostly good - a couple of minor issues noted. I'll also have to test this a bit later in the week once I'm back in the office.

File: image_view/package.xml
Comment (artineering) [reply image_view/package.xml]: I will fix that.

File: image_view/scripts/extract_images_sync
Comment (artineering) [reply image_view/scripts/extract_images_sync]: The original code used the parameter name with the underscore. Wasn't sure if I should change it. But, I can go ahead and drop the underscore.

**Comment by artineering** [comment]
I had inadvertently changed the api signature for cv_bridge.cvtColorForDisplay. Have reverted that change.

File: image_view/package.xml
Comment (ahcorde) [inline image_view/package.xml]:
```suggestion
  <exec_depend>rclpy</exec_depend>
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by artineering** [comment]
Thank you, @ahcorde for catching that mistake.

File: image_view/package.xml
Comment (artineering) [reply image_view/package.xml]: > This can be reverted - maintainer is different than contributor - it's an ongoing commitment (and usually involves commit rights, reviewing PRs and releasing updates)

What is the usual process for becoming a maintainer? There isn't much information about these processes anywhere to get a clear picture of what is needed to become part of a project.

Comment (mikeferguson) [reply image_view/package.xml]: A lot of smaller projects don't have an official policy - generally:

1) existing maintainers recruit other maintainers as needed, usually from people who have made consistent or significant contributions to the project over a period of time

OR

2) if all the maintainers abandon a project, within ROS, we have some admins for the ros-* orgs that can help on board new maintainers. 

Some larger projects (for instance, MoveIt) have a whole process for existing maintainers to actually vote on new promoting "core contributors" to "maintainers" (they keep a full list here: https://moveit.ros.org/about/)

I think the key to keep in mind though: ANYBODY can open a PR and work to get something into any ROS package. The response time from various maintainers obviously varies wildly. Contributing to a project is sort of a one-time commitment to get a particular PR over the threshold for acceptance, whereas maintainers are more of a longer-term commitment. With many ROS projects, you'll also find that many maintainers aren't even really actively developing on the project anymore, their contribution is in terms of 1) knowledge of the project, 2) time to review PRs, 3) periodically pushing new releases, and 4) often having access to hardware to test PRs on.

Comment (artineering) [reply image_view/package.xml]: > A lot of smaller projects don't have an official policy - generally:
> 
> 1. existing maintainers recruit other maintainers as needed, usually from people who have made consistent or significant contributions to the project over a period of time
> 
> OR
> 
> 2. if all the maintainers abandon a project, within ROS, we have some admins for the ros-* orgs that can help on board new maintainers.
> 
> Some larger projects (for instance, MoveIt) have a whole process for existing maintainers to actually vote on new promoting "core contributors" to "maintainers" (they keep a full list here: https://moveit.ros.org/about/)
> 
> I think the key to keep in mind though: ANYBODY can open a PR and work to get something into any ROS package. The response time from various maintainers obviously varies wildly. Contributing to a project is sort of a one-time commitment to get a particular PR over the threshold for acceptance, whereas maintainers are more of a longer-term commitment. With many ROS projects, you'll also find that many maintainers aren't even really actively developing on the project anymore, their contribution is in terms of 1) knowledge of the project, 2) time to review PRs, 3) periodically pushing new releases, and 4) often having access to hardware to test PRs on.

Ah, that makes sense. Thanks for sharing.

**Review by ahcorde — APPROVED** [review]

**Comment by mikeferguson** [comment]
3 notes from me:
 * Can you pop the last commit off of this? We'll end up squashing the history when we merge this, so I suggest we make a separate PR afterwards for the sec_per_frame update - that will make the history far more clear for anyone coming back and looking at those changes in the future.
 * I think we need to add an install for the script (I don't see anything in CMake - I presume "ros2 run" doesn't actually work? Unless there is some magic in the auto_ament commands that I'm not aware of)
 * Can you give a quick description of your "testing"? I didn't get a chance to try this out today - if you can describe how you tested it, that might give me the confidence to merge without actually testing it myself

**Comment by mikeferguson** [comment]
And one other question if you don't mind - how did you find out about these issues and decide to work on them? (this is more for calibrating my notions of "what gets people involved in ROS")

**Comment by artineering** [comment]
> 3 notes from me:
> 
> * Can you pop the last commit off of this? We'll end up squashing the history when we merge this, so I suggest we make a separate PR afterwards for the sec_per_frame update - that will make the history far more clear for anyone coming back and looking at those changes in the future.

_I agree. I did not intend to this change to get included in this. But, since I am fairly new to github, I didn't realize that submitting it to my fork would automatically update the PR. I will try to squash the change on my side and resubmit once this PR is merged._

> * I think we need to add an install for the script (I don't see anything in CMake - I presume "ros2 run" doesn't actually work? Unless there is some magic in the auto_ament commands that I'm not aware of)

_Yes, the script does not run by itself. In order to run it, you need to create a ros package and copy the script into it. Thats what I did to test the script upto the point where I was able to build and run. I tried out all ros arguments._

> * Can you give a quick description of your "testing"? I didn't get a chance to try this out today - if you can describe how you tested it, that might give me the confidence to merge without actually testing it myself

_I tested the build and run part of the script. I also tested whether all the ros arguments are getting picked up as expected. I couldn't get the camera/image_raw to publish properly the last time I ran the code in ros2, so that part is pending._

**Comment by artineering** [comment]
> And one other question if you don't mind - how did you find out about these issues and decide to work on them? (this is more for calibrating my notions of "what gets people involved in ROS")

My background is in Image Processing and have started my PhD in Computer Science recently. In discussions with my professor and other TAs, contributing to ROS was one of the things recommended to get ourselves familarized with the ecosystem. ROS-perception felt like the obvious choice for me. Your tag ('good first issue') helped identify the issues that I felt confident enough to handle.

I do plan to contribute to this repository as best I can by taking on more issues and hopefully contributing some new features as well.

**Comment by mikeferguson** [comment]
I just pushed a commit with the install directive and tested this locally - was able to save images from a Primesense camera with:

```
ros2 launch openni2_camera camera_with_cloud.launch.py
ros2 run image_view extract_images_sync --ros-args -p inputs:='[/camera/rgb/image_raw, /camera/depth/image_raw]' -p approximate_sync:=True
```

I did have to slightly patch the encoding stuff (cv_bridge says 8UC3 isn't valid) - I'll put that in a follow on PR once I look into it in more detail.

**Review by mikeferguson — APPROVED** [review]

**Pull request merged**
